File: code/ch08/8.1.4_155.min-stack.py
class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min = float('inf')

    def push(self, x: int) -> None:
        if not self.stack:
            self.min = x
            self.stack.append(0)
        else:
            self.stack.append(x - self.min)
            if x < self.min: self.min = x

    def pop(self) -> None:
        if not self.stack: return

        top = self.stack.pop()
        if top < 0:
            self.min = self.min - top

    def top(self) -> int:
        top = self.stack[-1]
        if top < 0: return self.min
        
        return top + self.min

    def getMin(self) -> int:
        return self.min

File: code/ch08/8.1.4_155.min-stack.cpp
class MinStack {
    typedef long long ll;
public:
    ll topElem; 
    /** initialize your data structure here. */
    MinStack() {
    }

    void push(int x) {
        if(s.empty()) {
            minn = x;
            s.push(x-minn);  //(当前值x-最小值minn)进栈
        }
        else {
            s.push(x-minn);
            minn = x < minn ? x : minn;  //值得注意的是当栈顶元素<0,则说明栈顶元素就是minn
        }
    }
    
    void pop() {
        if(!s.empty()) {
            topElem = s.top();
            s.pop();
            //原来的值x-最小值minn = topElem
            //topElem<0,说明原来值x更小(原来的值就是现在的minn)，则现在的minn应该被更新minn-topElem
            if(topElem < 0) minn -= topElem;
        }
        return;
    }
    
    int top() {
        if(!s.empty()) {
            topElem = s.top();
        }
        if(topElem < 0) return minn; //说明topElem <0, 说明 x<minn, minn已经被更新为当栈顶x，栈顶元素就是minn
        return topElem+minn;
    }
    
    int getMin() {
        return minn;
    }
    private:
        ll minn;
        stack<ll>s;
};